### ISSUE TITLE
 Random Access for FITS Files on S3

### ISSUE THREAD

**Issue by lhy-a** [nom-tam-fits/nom-tam-fits#707]
# Issue: Random Access for FITS Files on S3

## Problem
I need to read all the headers of a FITS file stored on S3. However, for large files, this process requires reading the entire file, which can consume a significant amount of time. 

The good news is that the `RandomAccessFileIO` interface allows random access to remote FITS files when implemented correctly. This would enable efficient header retrieval without downloading the entire file.

## Challenge
Most S3 SDKs provide `InputStream` implementations that do not fully support the methods required by the `RandomAccessFileIO` interface. From my testing, I found that even a partial implementation of the interface is sufficient for random access to FITS headers.

## Proposal
When I attempted to create my own `RandomAccessFileIO` wrapper for S3, I found very few resources available online. I believe adding a simple example implementation to the `README.md` file or documentation would be extremely helpful for others facing similar issues.

## Request
I would be happy to contribute my implementation as an example if it could assist others in the community. Please let me know if this would be a valuable addition.

```java
import java.io.FileDescriptor;
import java.io.IOException;
import java.nio.channels.FileChannel;

import nom.tam.fits.BasicHDU;
import nom.tam.fits.Fits;
import nom.tam.fits.Header;
import nom.tam.fits.HeaderCard;
import nom.tam.util.Cursor;
import nom.tam.util.RandomAccessFileIO;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.fs.FSDataInputStream;
import org.apache.hadoop.fs.FileSystem;
import org.apache.hadoop.fs.Path;

public class S3DataInputStreamAdapter implements RandomAccessFileIO {
  private final FSDataInputStream fsDataInputStream;
  private final long size;

  public S3DataInputStreamAdapter(FSDataInputStream fsDataInputStream, long size) {
    this.fsDataInputStream = fsDataInputStream;
    this.size = size;
  }

  @Override
  public String readUTF() throws IOException {
    return fsDataInputStream.readUTF();
  }

  @Override
  public FileChannel getChannel() {
    throw new UnsupportedOperationException("getChannel is not supported by FSDataInputStream");
  }

  @Override
  public FileDescriptor getFD() throws IOException {
    return fsDataInputStream.getFileDescriptor();
  }

  @Override
  public void setLength(long length) throws IOException {
    throw new UnsupportedOperationException("getChannel is not supported by FSDataInputStream");
  }

  @Override
  public void writeUTF(String s) throws IOException {
    throw new UnsupportedOperationException("writeUTF is not supported for FSDataInputStream");
  }

  @Override
  public void close() throws IOException {
    fsDataInputStream.close();
  }

  @Override
  public long position() throws IOException {
    return fsDataInputStream.getPos();
  }

  @Override
  public void position(long n) throws IOException {
    fsDataInputStream.seek(n);
  }

  @Override
  public long length() throws IOException {
    return size;
  }

  @Override
  public int read() throws IOException {
    return fsDataInputStream.read();
  }

  @Override
  public int read(byte[] b, int from, int length) throws IOException {
    return fsDataInputStream.read(b, from, length);
  }

  @Override
  public void write(int b) throws IOException {
    throw new UnsupportedOperationException("Write operations are not supported");
  }

  @Override
  public void write(byte[] b, int from, int length) throws IOException {
    throw new UnsupportedOperationException("Write operations are not supported");
  }

  public static void main(String[] args) throws IOException {
    Configuration conf = new Configuration();
    conf.set("fs.s3a.access.key", "xxxxxxx");
    conf.set("fs.s3a.secret.key", "xxxxxxx");
    conf.set("fs.s3a.endpoint", "http://xxxx.com");
    conf.set("fs.s3a.impl", "org.apache.hadoop.fs.s3a.S3AFileSystem");
    Path rootPath =
            new Path(
                    "s3a://your_bucket_name/your.fits");
    FileSystem fs = rootPath.getFileSystem(conf);
    FSDataInputStream fileInputStream = fs.open(rootPath);
    S3DataInputStreamAdapter adapter =
            new S3DataInputStreamAdapter(fileInputStream, fs.getFileStatus(rootPath).getLen());
    Fits fits = new Fits(adapter);
    BasicHDU<?> hdu;
    int hduIndex = 0;
    while ((hdu = fits.readHDU()) != null) {
      Header header = hdu.getHeader();
      if (header != null) {
        Cursor<String, HeaderCard> iter = header.iterator();
        while (iter.hasNext()) {
          HeaderCard card = iter.next();
          String key = "HEADER_" + hduIndex + "_" + card.getKey();
          String value = card.getValue() != null ? card.getValue() : "null";
          System.out.printf("key: %s, value: %s%n", key, value);
        }
      }
      hduIndex++;
    }
  }
}

```

**Comment by attipaci** [comment]
Hi Ihy-a,

I think an S3 adapter would make a good addition to the library. I myself have never used S3 so I don't have much relevant expertise in this area. However, @at88mph has. In fact, he is the one who contributed the streaming classes that can be used with S3 resources also. Given that CADC already has their own `RandomAccessFileIO` implementation for S3,  it would make sense to work with them and @at88mph to come up with a good adaptable design that can be generic enough to be used by everyone.

I'll be happy to review PRs, of course, and provide generic comments on naming / code structuring etc.

I hope @at88mph will join this thread, and that the two of you can hammer out the technical aspects.

**Comment by lhy-a** [comment]
> Hi @lhy-a,
> 
> I think an S3 adapter would make a good addition to the library. I myself have never used S3 so I don't have much relevant expertise in this area. However, @at88mph has. In fact, he is the one who contributed the streaming classes that can be used with S3 resources also. Given that CADC already has their own `RandomAccessFileIO` implementation for S3, it would make sense to work with them and @at88mph to come up with a good adaptable design that can be generic enough to be used by everyone.
> 
> I'll be happy to review PRs, of course, and provide generic comments on naming / code structuring etc.
> 
> I hope @at88mph will join this thread, and that the two of you can hammer out the technical aspects.

I’ve reviewed CADC’s implementation, and extracting the generic parts of the code should be quite straightforward for @at88mph. I’d be happy to assist in any way I can to help with this effort.

**Comment by attipaci** [comment]
Thanks @lhy-a,

Don't feel too contrained by the CADC implementation. If parts of it don't make sense to you, or you feel that you can do better, feel free to diverge for the best results. 

When you have a PR, I'll take a look. I'll probably suggest some refactoring and tweaks here and there. Just make sure you have test coverage for all of the new code (it should be 100% coverage for all additions, except for the lines that are quite impossible to reach).

Thanks for your efforts and good luck!

**Comment by robyww** [comment]
This is great! We need to do this very thing as well at IPAC with Firefly. I was about to start trying this in a few months so I am very glad for this work.

**Comment by at88mph** [comment]
Hi all.  There is certainly interest in this, which is great.  The CADC doesn't use S3 specifically, but rather CephFS.  However, I don't see this implementation necessarily containing S3 specific components, but rather something like an underlying `SeekableByteChannel` or something similar.  Most clients, such as the S3 Java one, should be able to produce one of those.  I'm experimenting with this now just to see how much work it would really be.

**Comment by attipaci** [comment]
OK, given the small crowd here of potential contributors, I'll suggest that @at88mph will put together an initial PR (if Dustin agrees to it, of course), hopefully soon (as in the next few weeks). Then @lhy-a and @robyww and I can test it out, tweak it, or make suggestions etc. To make it to the next release window, we have until late February to come up with something polished (otherwise it's late May next...)

If you feel too constrained with working on the same PR all at once, then feel to create your own independent or derivative PRs also. In the end we can discuss how we can merge the best ideas from each into a 'master' version. (It's more work for me, but I don't mind ;-). And of course, let me know if you have objections to this model...

**Comment by at88mph** [comment]
In order to prevent unnecessary dependencies from being included in the core library, I was thinking it may be better to create a separate library.  I have a working version here:
https://github.com/at88mph/nom-tam-fits-s3

The code coverage isn't at the desired level, but would the layout work?  Rather than a seekable stream, the implementation uses the `S3Client` to send byte range requests and pull down only the necessary bytes.  I'm not sure if there is a way to cache this or not if the `Fits` class wants to iterate over the same set of bytes.

If this is the incorrect approach, there is this library:
https://github.com/awslabs/aws-java-nio-spi-for-s3

which appears to be actively maintained.

**Comment by attipaci** [comment]
Hi @at88mph,

I lije the idea of a separate add-on library. However, it should not be based on a fork of nom-tam-fits, in my opinion, but rather have nom-tam-fits as a build and runtime maven dependence, so as to avoid duplicating code and future divergence. This way if someone want to use S3 with nom-tam-fits, they would install _BOTH_ the nom-tam-fits and the S3 add-on libraries as needed...

This, the S3 'module' should only contain the `RandomAccessFileIO` implementation, and have nom-tam-fits (and other libraries) as its dependency....

Does it make sense?

**Comment by at88mph** [comment]
I agree and that's exactly what this would be.  It's confusing because I copied over the user guide and the documentation, sorry about that.  Otherwise, this would be a very lean additional dependency.

**Comment by attipaci** [comment]
Sounds good. It just shows that I did not dive deep (shame on me!), and when I saw the copied user's guide I just assumed it was a fork without looking further. Mea culpa. Keep on going with the good work! Once it is ready for showtime, I'll reference the repo in the nom-tam-fits README (and the Maven package also, if there is going to be one).

**Comment by lhy-a** [comment]
> In order to prevent unnecessary dependencies from being included in the core library, I was thinking it may be better to create a separate library. I have a working version here: https://github.com/at88mph/nom-tam-fits-s3
> 
> The code coverage isn't at the desired level, but would the layout work? Rather than a seekable stream, the implementation uses the `S3Client` to send byte range requests and pull down only the necessary bytes. I'm not sure if there is a way to cache this or not if the `Fits` class wants to iterate over the same set of bytes.
> 
> If this is the incorrect approach, there is this library: https://github.com/awslabs/aws-java-nio-spi-for-s3
> 
> which appears to be actively maintained.

I noticed that the usage of aws-java-nio-spi-for-s3 is non-intrusive, and unlike the standard S3Client, it doesn’t require establishing a new HTTP connection for each random access operation.

I have submitted an example of an adapter implementation using aws-java-nio-spi-for-s3. It may require further modifications or encapsulation to meet specific requirements. I hope this could be helpful. @at88mph @attipaci

**Comment by at88mph** [comment]
Thanks @lhy-a, I will take a look.

**Comment by at88mph** [comment]
I think I like the approach of using the `FileChannel` in the end.  I'm trying to come up with a decent Unit Test and fix the `README`.  It does use the `S3AsyncClient` ultimately, so new HTTP connections will be made when necessary, although I think it's better at caching bytes locally and doing some smarter read-ahead operations.

For some background, @attipaci, this dependency will configure a new `FileSystemProvider`, and it looks like it just creates handling for `Path`s created from `s3` or `s3x` `URI`s.  Meaning, if you provide it a basic file `Path`, such as `/tmp`, it will still use the default `FileSystemProvider`.  This has not been tested, however.

All S3 specific configuration, such as authentication and how buckets are treated are handled by System Properties or Environment variables, but that will all be the responsibility of the caller.  Any thoughts?  The `nom-tam-fits-s3` library will be small and only contain the main implementation class and a unit test likely.

**Comment by at88mph** [comment]
Actually, looking closer at this, there is nothing S3 specific about it, other than including the dependency.  I wonder if we should simply add this new `FileChannelRandomAccessFileIO` to the core library, then add documentation on how to enable S3 in their application.  This would add a generic `FileChannel` implementation of the `RandomAccessFileIO`, with no further dependencies.

**Comment by attipaci** [comment]
I like where it's going... See if you can put together a PR for it with tests. 

I think there is still room for a separate S3 add-on package  to add S3 support out of the box with some nice configuration API for using it. And other add-on packages may add support for other services the same way... 

But, let's get the adapter class/interface into nom-tam-fits first. Maybe we can shorten the name to `RandomFileChannelIO`... 

Then perhaps @lhy-a might be interested in creating a super-simple add-on package specifically for S3 (with the additional specific dependency, and some configuration API to make it super easy to use...) -- if such a thing makes sense still...

**Comment by lhy-a** [comment]
> I like where it's going... See if you can put together a PR for it with tests.
> 
> I think there is still room for a separate S3 add-on package to add S3 support out of the box with some nice configuration API for using it. And other add-on packages may add support for other services the same way...
> 
> But, let's get the adapter class/interface into nom-tam-fits first. Maybe we can shorten the name to `RandomFileChannelIO`...
> 
> Then perhaps @lhy-a might be interested in creating a super-simple add-on package specifically for S3 (with the additional specific dependency, and some configuration API to make it super easy to use...) -- if such a thing makes sense still...


I agree with @at88mph’s idea, and as a result, a dedicated nom-tam-fits-s3 library with S3 dependencies is no longer necessary.
	1.	Thanks to the existence of the S3 SPI,  nom-tam-fits-s3 only needs to implement an adapter for FileChannelRandomAccessFileIO (or PathRandomAccessFileIO).
	2.	For end-users, the README or documentation should include a guide on how to use it. After all, users need to include the S3 SPI dependency themselves, and not everyone knows about SPI. Honestly, I didn’t either, but now I see it’s a really useful mechanism.
	3.	For developers, if we want to support other third-party cloud services in the future besides S3, we (or others) just need to develop a new SPI implementation.

As for my personal situation, the issue I faced has already been resolved with the example I initially provided. However, a community-maintained solution would undoubtedly be more robust.

**Comment by at88mph** [comment]
Added PR #717
